src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            return new ForeachInfo(countName, explicitCast, requireCollectionStatement, foreachStatement);
+        }
+
+        private static void GetInterfaceInfo(
+            ISemanticFactsService semanticFact, SemanticModel model, ILocalSymbol foreachVariable, IOperation foreachCollection, out string explicitCast, out string countName)
+        {
+            explicitCast = default;
+            countName = null;
+
+            // go through list of types to find out right set;
+            var foreachType = foreachVariable.Type;
+            if (IsNullOrErrorType(foreachType))
+            {
+                return;
+            }
+
+            // check array case first.
+            var collectionType = foreachCollection.Type;
+            if (IsNullOrErrorType(collectionType))
+            {
+                return;
+            }
+
+            if (collectionType is IArrayTypeSymbol array &&
+                array.Rank == 1 &&
+                semanticFact.IsAssignableTo(array.ElementType, foreachType, model.Compilation))
+            {
+                explicitCast = null;
+                countName = Length;
+                return;
+            }
+
+            // check ImmutableArray case second
+            if (collectionType.OriginalDefinition.Equals(model.Compilation.GetTypeByMetadataName(typeof(ImmutableArray<>).FullName)))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCast = null;
+                    countName = Length;
+                    return;
+                }
+            }
+
+            // go through all known interfaces we support
+            var knownCollectionInterfaces = s_KnownInterfaceNames.Select(
+                s => model.Compilation.GetTypeByMetadataName(s)).WhereNotNull().Where(t => !IsNullOrErrorType(t));
+
+            // check type itself is interface case
+            if (collectionType.TypeKind == TypeKind.Interface && knownCollectionInterfaces.Contains(collectionType.OriginalDefinition))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+            }
+
+            // check regular cases (implicitly implemented)
+            string explicitInterface = null;
+            foreach (var current in collectionType.AllInterfaces)
+            {
+                if (!knownCollectionInterfaces.Contains(current.OriginalDefinition))
+                {
+                    continue;
+                }
+
+                // see how the type implements the interface
```

Comment:
This seems super baroque.  Is this really necessary.  If the type implmeents both IList and is IEnumerable, it seems reasonable to assume that they're not doing weird things, and that you can switch from enumerating to indexing safely.

Reply:
I am not sure why checking assignability is bad. if IEnumerable<string> and IList<int> is implemented, I dont see point why we offer refactoring.

if IEnumerable<string> and IList<string> is implemented, then regardless of the checking, refactoring will be offered.

Reply:
Because foreach will blindly cast.  The type taht is being enumerated does not have to be assignable to the foreach variable.  For example, this is legal:

```c#
    void Foo(IList<object> list)
    {
        foreach (string s in list)
        {

        }
    }
```

It seems like your feature would not work here.